rfcs/0004-categorized-attributes.md

```diff
@@ -0,0 +1,217 @@
+
+A parser MUST abstract the presence of _old-form_ _categorized attributes_ from the engine.
+
+### Generators
+
+A generator for the save file format MUST support _new-form_ attribute generation. It SHALL NOT generate _old-form_ syntax for _categorized attributes_. It MUST NOT generate duplicate entries of the _category_, _effect_, _modifier_ or combined _modifier_ + _effect_ keywords.
+
+A generator SHALL use a data node to represent a _category_ and _effect_ or _effect_ + _category_ combination. A generator SHOULD NOT include attributes with their default value.
+
+A generator SHOULD use a deterministic output ordering for attributes, and MAY order them alphabetically. It SHALL provide the most concise output possible.
+
+A generator MAY NOT support _old-form_ _categorized attributes_.
+
+### Engine support
+
+The supported _categories_ SHOULD include all of the following:
+
+```
+shield generation
+hull repair
+thrusting
+reverse thrusting
+steering
+active cooling
+ramscooping
+cloaking
+afterburning
+firing
+```
+
+It MAY include any or all of the following:
+```
+protection
+resistance
+damage
+capacity
+```
+
+The supported _effects_ SHOULD include all of the following:
+```
+shields
+hull
+thrust
+reverse thrust
+turn
+active cooling
+ramscoop
+cloak
+cooling
+force
+energy
+fuel
+heat
+jam
+```
+
+It MAY include any or all of the following:
+```
+disabled
+minable
+piercing
+```
+
+The supported _modifiers_ SHOULD include all of the following:
+```
+none
+multiplier
+relative
+over time
+```
+
+The _default modifier_ MUST be `none`.
+
+### Implementation
+
+The constraints of attribute values MUST be enforced in runtime. _Categorized attributes_ with their default value MAY NOT be stored. There MUST exist an abstraction layer supporting queries and updates of stored and non-stored _categorized attributes_. This abstraction layer SHOULD support non-_categorized attributes_ as well. This layer MUST support storing any _category_+_effect_+_modifier_ combination as a _categorized attribute_, regardless of engine support for such an attribute.
+
+Querying the value of a non-stored attribute MUST return its default value, and MUST NOT cause exceptions or undefined behaviour. 
+
+An implementation MUST allow efficient query of all _categorized attributes_ when given a _category_. It MUST allow for efficient transformation of all _categorized attributes_ of a given category, such as adding another set of matching attributes to it on a per-value basis.
+
+For instance, such an operation MUST be performed efficiently:
```

Comment:
Yes, this is the in-engine benefit that we are looking for.
It is not directly 1 to 1 related to the keywords though, we can still remap some things during load and save if we want to.